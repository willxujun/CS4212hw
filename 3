//chained arithmetic, relational operations
class Main {
    Box b;

    Void main() {
        Box b;
        Int sum;
        Bool b1;
        Bool b2;
        Int i1;
        Int i2;

        if(!(!b1 && (b2 || b2))) {
            sum = (i1 + i2 -i2) /i1 * i2;
        } else {
            return;
        }
        while(i1 + i2 <= -i1) {
            b1 = b1 && b2 && true;
            readln(b2);
            println(i1);
        }
        b = new Box();

        b.x = 10;
        b.y = 20;
        b.z = 30;
        b.x = b.x + b.y;
        this.b.x = 10;
        this.b.y = 20;
        this.b.z = 30;

        sum = this.b.func("Error", 1000);
        println(sum);
    }
}

class Box {
    Int x;
    Int y;
    Int z;
    Box b1;

    Int func(String s1, Int one) {
        Int sum;
        sum = x + y + z + one;
        println(s1);
        return sum;
    }
}
