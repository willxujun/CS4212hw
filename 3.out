Error at line 16, column 44 : not an ascii character <\xff>
Error at line 16, column 49 : not an ascii character <\999>
Error: l expression has different type from r
In method main
In class Main

Error: Assigning type Bool to t1 of type Int
In method main
In class Main

Error: Constructor of unknown class Compute
In method main
In class Main

Error: r value does not type check in assignment to help
In method main
In class Main

Error: field assign: RHS of l value does not type check for field chachedValue
In method main
In class Main

Error: Unknown function binding atom
In method main
In class Main

Error: Calling a non-function type
In method main
In class Main

Error: field assign: LHS of l value does not type check for field id
In method main
In class Main

Unknwon Class Compute or unknown function addSquares
Error: Call to unknown function addSquares
In method main
In class Main

Error: Calling a non-function type
In method main
In class Main

Unknwon Class Compute or unknown function square
Error: Call to unknown function square
In method main
In class Main

Error: Calling a non-function type
In method main
In class Main

Unknwon Class Compute or unknown function square
Error: Call to unknown function square
In method main
In class Main

Error: Calling a non-function type
In method main
In class Main

Error: r value does not type check in assignment to t2
In method main
In class Main

Error: Should not reach here...
In method main
In class Main

Error: Calling a non-function type
In method main
In class Main

Error: Unknown variable binding dynamic
In method main
In class Main

Error: Dispatch object does not type check for call to dispatch
In method main
In class Main

Error: Calling a non-function type
In method main
In class Main

Error: Unknown variable binding id
In method main
In class Main

Error: Access RHS cannot find variable fieldname1
In method main
In class Main

Error: Return statement does not have appropriate type
In method main
In class Main

Program
  Class Main
    VariableList
    MethodList
      Void main
        FormalList
          Int i
          Bool a
          String b
          Void d
        VariableList
          Int t1
          Bool t2
          String t3
          Void t4
          Compute help
        StatementList
          If
            BinExp GT
              (Id t2)
              (Id t1)
            StatementList
              Print
                (Str (Square of uToÿ ϧ d larger than sum of squares))
            StatementList
              Print
                (Str (Square of d larger than sum of squares))
          While
            (Bool true)
            StatementList
              Assign t1
                (Id t2)
          Assign help
            New Compute
          Assign chachedValue
            (Id help)
            BinExp Mult
              (Id t1)
              (Int 3)
          Assign id
            Dispatch
              (Id atom)
              ExpressionList
                (Id exp1)
            (Id exp2)
          Assign t1
            BinExp Plus
              Dispatch
                Access addSquares
                  (Id help)
                ExpressionList
                  (Id a)
                  (Id b)
              Dispatch
                Access square
                  (Id help)
                ExpressionList
                  (Id i)
          Assign t2
            Dispatch
              Access square
                (Id help)
              ExpressionList
                (Id d)
          Dispatch
            Dispatch
              (Id staticmtd)
              ExpressionList
                (Id arg1)
            ExpressionList
              (Int 123)
          Dispatch
            Access dispatch
              (Id dynamic)
            ExpressionList
              (Id b)
          Read id
          Print
            Access fieldname1
              (Id a)
          Return
            (Int 1)
          Return
  Class Dummy
    VariableList
      Compute c
      Int i
      Dummy j
    MethodList
      Int dummy
        FormalList
        VariableList
          Bool i
          Bool j
        StatementList
          If
            BinExp Or
              (Id i)
              (Id j)
            StatementList
              Return
                (Int 1)
            StatementList
              While
                (Id i)
                StatementList
                  Assign i
                    UExp Not
                      (Id j)
              Assign c
                Dispatch
                  Access getCompute
                    This
                  ExpressionList
          Return
            Dispatch
              Access square
                Dispatch
                  Access getCompute
                    This
                  ExpressionList
              ExpressionList
                UExp Neg
                  (Int 3)
          Return
            (Id i)
      Compute getCompute
        FormalList
        VariableList
        StatementList
          Return
            (Id c)

Semantics error detected. Refer to msgs and parse tree.
