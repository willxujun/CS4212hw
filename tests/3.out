===== CData3 ===== 
class Box
    Int x
    Box b1
    Int z
    Int y
class Main
    Box b

===== CMtd3 ===== 
Void %Main_0 [Main this]
[Box b, Int sum, Bool b1, Bool b2, Int i1, Int i2, Bool _t0, Bool _t1, Bool _t2, Bool _t3, Int _t4, Int _t5, Int _t6, Int _t7, Int _t8, Int _t9, Bool _t10, Bool _t11, Bool _t12, Box _t13, Int _t14, Int _t15, Int _t16, Box _t17, Int _t18, Box _t19, Int _t20, Box _t21, Int _t22, Box _t23, Int _t24]
    _t0 = ! b1;
    _t1 = b2 || b2;
    _t2 = _t0 && _t1;
    _t3 = ! _t2;
    if _t3 goto 0;
    goto 1;
    L0:
    _t4 = i1 + i2;
    _t5 = _t4 - i2;
    _t6 = _t5 / i1;
    _t7 = _t6 * i2;
    sum = _t7;
    L1:
    return;
    L2:
    _t8 = i1 + i2;
    _t9 = - i1;
    _t10 = _t8 <= _t9;
    if _t10 goto 3;
    goto 4;
    L3:
    _t11 = b1 && b2;
    _t12 = _t11 && true;
    b1 = _t12;
    readln b2;
    println i1;
    goto 2;
    L4:
    _t13 = new Box();
    b = _t13;
    _t14 = b.x;
    _t14 = 10;
    _t15 = b.y;
    _t15 = 20;
    _t16 = b.z;
    _t16 = 30;
    _t17 = this.b;
    _t18 = _t17.x;
    _t18 = 10;
    _t19 = this.b;
    _t20 = _t19.y;
    _t20 = 20;
    _t21 = this.b;
    _t22 = _t21.z;
    _t22 = 30;
    _t23 = this.b;
    _t24 = %Box_0 [_t23, "Error", 1000];
    sum = _t24;
    println sum;

Int %Box_0 [Box this, String s1, Int one]
[Int sum, Int _t25, Int _t26, Int _t27]
    _t25 = x + y;
    _t26 = _t25 + z;
    _t27 = _t26 + one;
    sum = _t27;
    println s1;
    return sum;


